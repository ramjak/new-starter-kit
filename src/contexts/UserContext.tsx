import {
  createContext,
  useCallback,
  useContext,
  useMemo,
  useState,
} from "react";
import * as React from "react";
import { interfaces } from "inversify";
import realContainer from "../inversify.config";
import IAuthService from "../services/IAuthService";
import TYPES from "../services/types";

export interface IUser {
  username: string;
  token: string;
  email: string;
}

const defaultUserValue = {
  username: "",
  token: "",
  email: "",
};

export interface IUserContextValue {
  userData: IUser;
  login(user: IUser): void;
  logout(): void;
}

export const UserContext = createContext<IUserContextValue>({
  userData: defaultUserValue,
  login(): void {
    console.log("login");
  },
  logout(): void {
    console.log("logout");
  },
});

export const useUserContext = () => useContext(UserContext);

interface IUserContextProvider {
  container?: interfaces.Container;
}

export const UserContextProvider: React.FC<IUserContextProvider> = (props) => {
  const { children, container } = { container: realContainer, ...props };
  const authService = container.get<IAuthService>(TYPES.AuthService);
  const currentUser = authService.getAuthData();

  const [userData, setUserData] = useState<IUser>(
    currentUser || defaultUserValue
  );

  const login = useCallback(
    (user) => {
      authService.login(user).catch((err) => {
        throw err;
      });
      setUserData(user);
    },
    [authService]
  );

  const logout = useCallback(() => {
    authService.logout().catch((err) => {
      throw err;
    });
    setUserData({ token: "", username: "", email: "" });
  }, [authService]);

  const value = useMemo(() => ({ userData, login, logout }), [
    login,
    logout,
    userData,
  ]);

  return (
    <UserContext.Provider value={value}>
      {typeof children === "function" ? children(value) : children}
    </UserContext.Provider>
  );
};
