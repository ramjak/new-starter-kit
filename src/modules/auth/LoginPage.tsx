import { Button, Input } from "@material-ui/core";
import React, { ChangeEvent, FormEvent, useCallback, useState } from "react";
import ROUTES, { useNavigateTo } from "../../routes";
import { useUserContext } from "../../contexts/UserContext";

interface ILoginPage {}

interface ICredential {
  username: string;
  password: string;
}

const LoginPage = ({}: ILoginPage) => {
  const initValue: ICredential = { password: "", username: "" };
  const [value, setValue] = useState<ICredential>(initValue);
  const navigateTo = useNavigateTo();
  const { login } = useUserContext();

  const submitLogin = useCallback(
    (e: FormEvent<HTMLFormElement>) => {
      e.preventDefault();
      const token = `${value.username}-${value.password}`;
      login({
        token,
        username: value.username,
        email: `${value.username}@email.com`,
      });
      navigateTo(ROUTES.home);
    },
    [login, navigateTo, value.password, value.username]
  );

  const onChangeUsername = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setValue((state) => ({ ...state, username: event.target.value }));
    },
    []
  );

  const onChangePassword = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setValue((state) => ({ ...state, password: event.target.value }));
    },
    []
  );

  return (
    <form onSubmit={submitLogin}>
      <Input
        placeholder="Username"
        fullWidth={true}
        value={value.username}
        onChange={onChangeUsername}
      />{" "}
      <br />
      <Input
        type="password"
        placeholder="Password"
        fullWidth={true}
        value={value.password}
        onChange={onChangePassword}
      />{" "}
      <br />
      <Button type="submit" variant="outlined" fullWidth={true} color="primary">
        Log in
      </Button>
    </form>
  );
};

export default LoginPage;
