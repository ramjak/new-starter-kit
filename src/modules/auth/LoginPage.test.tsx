import "reflect-metadata";
import { createRenderer } from "react-test-renderer/shallow";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginPage from "./LoginPage";
import { UserContext } from "../../contexts/UserContext";
import ROUTES from "../../routes";

describe("LoginPage", () => {
  it("render correctly at first load", () => {
    const shallowRenderer = createRenderer();

    shallowRenderer.render(<LoginPage />);
    const result = shallowRenderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });

  it("process input correctly", () => {
    const navigateToMock = jest.fn();
    const useNavigateToMock = () => navigateToMock;
    const loginMock = jest.fn();
    const logoutMock = jest.fn();

    const { getByRole, getByPlaceholderText } = render(
      <UserContext.Provider
        value={{
          userData: { username: "", email: "", token: "" },
          login: loginMock,
          logout: logoutMock,
        }}
      >
        {/* eslint-disable-next-line react/jsx-no-bind */}
        <LoginPage useNavigateTo={useNavigateToMock} />
      </UserContext.Provider>
    );

    userEvent.type(getByPlaceholderText("Username"), "username123");
    expect(getByPlaceholderText("Username")).toHaveValue("username123");

    userEvent.type(getByPlaceholderText("Password"), "password");
    expect(getByPlaceholderText("Password")).toHaveValue("password");

    userEvent.click(getByRole("button", { name: /Log in/i }));

    expect(loginMock).toBeCalledWith({
      email: "username123@email.com",
      token: "username123-password",
      username: "username123",
    });
    expect(navigateToMock).toBeCalledWith(ROUTES.home);
  });
});
