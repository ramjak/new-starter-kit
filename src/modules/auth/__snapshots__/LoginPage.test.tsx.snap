// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage render correctly at first load 1`] = `
<form
  onSubmit={[Function]}
>
  <WithStyles(ForwardRef(Input))
    fullWidth={true}
    onChange={[Function]}
    placeholder="Username"
    value=""
  />
   
  <br />
  <WithStyles(ForwardRef(Input))
    fullWidth={true}
    onChange={[Function]}
    placeholder="Password"
    type="password"
    value=""
  />
   
  <br />
  <WithStyles(ForwardRef(Button))
    color="primary"
    fullWidth={true}
    name="submit"
    type="submit"
    variant="outlined"
  >
    Log in
  </WithStyles(ForwardRef(Button))>
</form>
`;
